Add support for `rebase rojig://`

cgwalters opened the pull request:
This is an initial drop of support for:
`rpm-ostree rebase rojig://fahc:fedora-atomic-host`.  We also
then support `rpm-ostree upgrade` from that.

There's a lot that could be improved here; the test coverage is relatively
minimal. A blocking issue there is having a realistic jigdo setup, and that's
going to require changing how we do testing. For now, this means that if we want
to e.g. change the format we'll have to temporarily disable this test, get the
format change in, update FAHC, then re-enable the test.

---

```
# rpm-ostree status
● ostree://fedora-atomic:fedora/27/x86_64/atomic-host
                   Version: 27.16 (2017-11-28 23:08:35)
                    Commit: 86727cdbc928b7f7dd0e32f62d3b973a8395d61e0ff751cfea7cc0bc5222142f
              GPGSignature: Valid signature by 860E19B0AFA800A1751881A6F55E7430F5282EE4
                  Unlocked: development
# cat /etc/yum.repos.d/local.repo 
[local]
baseurl=http://192.168.121.1:8000/f27-jigdo-test
gpgcheck=0
# rpm-ostree rebase --ex-jigdo local:fedora-atomic-host
Enabled rpm-md repositories: local

Updating metadata for 'local': [=============] 100%
rpm-md repo 'local'; generated: 2017-12-21 22:25:22
oirpm: fedora-atomic-host-27.26-1.fc27.x86_64 (local) commit=5e467248abd238bfcd9b6c49f34ecae3b1c3e8831b76ae7fad173eabe83cdd8b
457 packages to import, download size: 322.9 MB
Will download: 457 packages (322.9 MB)
Importing (457/457) [=============] 100%
# rpm-ostree status
State: idle
Deployments:
  jigdo://local:fedora-atomic-host
                   Version: 27.26 (2017-12-21 22:19:00)
                    Commit: 5e467248abd238bfcd9b6c49f34ecae3b1c3e8831b76ae7fad173eabe83cdd8b

● ostree://fedora-atomic:fedora/27/x86_64/atomic-host
                   Version: 27.16 (2017-11-28 23:08:35)
                    Commit: 86727cdbc928b7f7dd0e32f62d3b973a8395d61e0ff751cfea7cc0bc5222142f
              GPGSignature: Valid signature by 860E19B0AFA800A1751881A6F55E7430F5282EE4
                  Unlocked: development
```

cgwalters commented:
Split out the prep into https://github.com/projectatomic/rpm-ostree/pull/1168

cgwalters commented:
One thing I'm blocking on a bit here is deciding how we teach the whole codebase about this.  It's going to leak into the CLI and DBus etc.  I'm tentatively thinking of standardizing on the so-far tentative `ostree://` and `jigdo://` schema.  Basically then, everywhere we refer to a "refspec" it could be either one of those.

Hence then it'd be e.g.
```
# rpm-ostree rebase jigdo://local:fedora-atomic-host
```

dustymabe commented:
`jigdo://local:fedora-atomic-host` is looking like a URI. should we consider the way it looks since it's essentially a re-implementation of jigdo? i.e. should the URI be more specific since it's essentially rpm-ostree specific.

cgwalters commented:
For people who are already familiar with the original jigdo project, I agree we're introducing confusion.  http://lists.rpm.org/pipermail/rpm-ecosystem/2018-January/000532.html

The set of people familiar with jigdo already is fairly small though, and my initial opinion was that context is going to be sufficient to disambiguate, and where it isn't we'd generally type "rpm-ostree jigdo" (e.g. in docs).

That said I'm not fixed on that...we could for example use the term `rjigdo` or something?

cgwalters commented:
`rojig` and `rojigdo` aren't taken...the latter feels ugly, I could certainly warm to the former.

cgwalters commented:
Hmm, `rpmostree+jigdo://` isn't *that* bad actually...slightly verbose but precedent is e.g. `git+ssh://`.

jlebon commented:
Hmm, it is true that re-using the jigdo name might introduce confusion. I'd prefer `rojig://` over `rpmostree+jigdo://` I think.

An idea 💡: if the scheme represents the source from which we receive updates, then maybe we should be focusing on the fact that we're using an ostree repo in one and a yum repo in the other. So e.g. something like `ostree://` and `yum://` ? (And let jigdo be an implementation detail). That also lines up nicely with the `/etc` dir names where the remote/repo configs are stored.

/bikeshed :)

cgwalters commented:
Hmm...feels a bit like using `yum://` here would add even more confusion - with "pure jigdo" mode we're only using like ~15% of the libdnf side.  We're really *not* doing package management.

Of course though you're absolutely right that `/etc/yum.repos.d` is part (a critical one actually) of that ~15%...

cgwalters commented:
Let's try to resolve this relatively soon, as the naming is going to get *very* baked into things.

cgwalters commented:
I sent mail to the [original jigdo](http://atterer.org/jigdo/) maintainer FWIW.

jlebon commented:
Hmm, I think I still have a slight preference for `yum://` given that it makes things really easy to explain: "`ostree://` origins use ostree remotes from `/etc/ostree/remotes.d` and `yum://` origins use yum repos from `/etc/yum.repos.d`". The downside is that if we somehow ever come up with another method of distribution that also uses yum repos in some way, then things can get confusing. Re. dependency resolution: that's a good point, though I wonder how many people would automatically make that leap upon seeing "`yum://`".

In the end we'll need some solid docs for this, so either `yum://` or `rojig://` sound good to me!

cgwalters commented:
In this discussion are we talking about rebranding the whole thing (e.g. it'd be `--ex-yum-output-set` instead of `--ex-jigdo-output-set`, and all of our docs would change) or just the URI?

jlebon commented:
I was thinking more just the URI. On the client side, there's no other place where jigdo "leaks out", right?

cgwalters commented:
I'm handwaving here but I think users are going to fall into either "technical" or "not technical".  For technical users I'm very doubtful they'll not at some point have to understand jigdo.  Or to avoid a double negative: I think technical users are going to eventually need to understand jigdo.  

For nontechnical users I'd like to maintain some ability to Google search for problems; the `yum` term is problematic there right?

Maybe this is all arguing for `rojig` or `rojigdo`.

jlebon commented:
> For nontechnical users I'd like to maintain some ability to Google search for problems

Ahh, very good point, I hadn't thought about that. Esp. given that this will be directly exposed to users. From that same argument, let's go with `rojig`, then? Pays tribute to the idea, but still unique to rpm-ostree. (I mean "the whole thing" here, not just the URI).

cgwalters commented:
@jlebon I notice we're both modifying `status.c`...any thoughts on which to land first?

cgwalters commented:
To clarify: Is `rojig` "short" for "rpm-ostree jigdo"?  Would we still talk about "jigdo" in rpm-ostree context or are you really thinking we'll replace the term "jigdo" entirely in the source/text etc.?

dustymabe commented:
> One thing I'm blocking on a bit here is deciding how we teach the whole codebase about this. It's going to leak into the CLI and DBus etc. I'm tentatively thinking of standardizing on the so-far tentative ostree:// and jigdo:// schema.

dumb question here: why don't we store the repo type (jigdo/ostree) along with the remote configuration?

cgwalters commented:
We're not using libostree remotes in jigdo mode; libostree is not involved in doing HTTP at all.  You may be thinking of the origin file?  If so that's exactly what this PR is doing.

cgwalters commented:
A simple way to think of jigdo mode is that the role of libostree is a *lot* closer architecturally to e.g. LVM.

dustymabe commented:
ehh. i think I'm more confused - mainly because I don't know the internals. Let me see if I can put it in other terms. for an `ostree:///` remote I do something like this in the past:

```
# ostree remote add f22-test http://192.168.122.219:43620 --no-gpg-verify
# rpm-ostree rebase f22-test:fedora-atomic/f22/x86_64/docker-host
```

In the jigdo world, what does this look like?

cgwalters commented:
@dustymabe That's discussed in https://github.com/projectatomic/rpm-ostree/issues/1081#issuecomment-348540604 and below.  TL;DR the rpm-md repo takes over the role of ostree remote.

cgwalters commented:
Also see the [first comment](https://github.com/projectatomic/rpm-ostree/pull/1166#issue-284023234) here - note the `local` repo which is an rpm-md/yum/dnf repo.

dustymabe commented:
> @dustymabe That's discussed in #1081 (comment) and below. TL;DR the rpm-md repo takes over the role of ostree remote.

ok it's starting to make a little more sense, although I think I'd like to discuss the design a bit more at some point.

so given the current information there are ostree remotes and yum repo remotes, can we not assume ostree remotes imply `ostree://` and yum repo remotes imply `jigdo://`?

jlebon commented:
> @jlebon I notice we're both modifying status.c...any thoughts on which to land first?

It shouldn't be too hard to rebase on yours. Though I do have https://github.com/projectatomic/rpm-ostree/pull/1196 ready, which is the majority of my work in `status.c` (I'm thinking of putting the cached update printing in `libbuiltin` for easier sharing with `upgrade --check/--preview`).

> To clarify: Is `rojig` "short" for "rpm-ostree jigdo"? Would we still talk about "jigdo" in rpm-ostree context or are you really thinking we'll replace the term "jigdo" entirely in the source/text etc.?

I guess I mean at least in places that are user facing on the client/server sides? (I.e. switches/outputs/etc...).

jlebon commented:
> I guess I mean at least in places that are user facing on the client/server sides? (I.e. switches/outputs/etc...).

OTOH, all those are still under `ex` labels, so we can leave them as they are for now? (And just change the URI).

cgwalters commented:
Going to leave this untouched for a little bit and peel off some prep commits.

cgwalters commented:
Rebased, updated, now with basic test.  Lifting WIP.

jlebon commented on src/daemon/rpmostreed-transaction-types.c:
```diff
@@ -48,15 +48,34 @@ change_origin_refspec (OstreeSysroot *sysroot,
+   * `rpm-ostree rebase updates:fedora-atomic-workstation` etc.
+   */
+  if (refspectype == RPMOSTREE_REFSPEC_TYPE_ROJIG)
+    {
```
Let's add a check that they're not the same? Like we do in the ostree case.

jlebon commented on tests/vmcheck/test-jigdo-client.sh:
```diff
@@ -0,0 +1,40 @@
+gpgcheck=0
+EOF
+
+vm_rpmostree rebase rojig://fahc:fedora-atomic-host
```
Can we require a `--ex-rojig` option here to explicitly mark this as experimental for now?

jlebon commented on src/daemon/rpmostreed-deployment-utils.c:
```diff
@@ -295,31 +297,32 @@ rpmostreed_deployment_generate_variant (OstreeSysroot *sysroot,
-    return NULL;
-
-  if (pending_base_commitrev && strcmp (pending_base_commitrev, base_checksum) != 0)
+  if (refspec && refspec_type == RPMOSTREE_REFSPEC_TYPE_OSTREE)
```
Why gate on `refspec` here? We should always have one, right? In fact, the origin errors out right now if there isn't one.

Similarly down below where we insert the `origin` key.

jlebon commented on src/daemon/rpmostree-sysroot-upgrader.c:
```diff
@@ -374,58 +374,108 @@ rpmostree_sysroot_upgrader_pull_base (RpmOstreeSysrootUpgrader  *self,
+        g_key_file_set_string (tsk, "tree", "jigdo", refspec);
+        const char *jigdo_version = rpmostree_origin_get_jigdo_version (self->origin);
+        if (jigdo_version)
+          g_key_file_set_string (tsk, "tree", "jigdo-version", jigdo_version);
```
The `jigdo-version` stuff isn't hooked up yet on the `DeployTransaction` side, right? E.g. one can't yet do `rpm-ostree rebase updates:$jigdorpm $version`, right?

cgwalters replied:
Right, not yet.

cgwalters commented on src/daemon/rpmostreed-transaction-types.c:
```diff
@@ -48,15 +48,34 @@ change_origin_refspec (OstreeSysroot *sysroot,
+      *out_new_refspec = g_strdup (src_refspec);
+      return TRUE;
+    }
 
```
Converted to a `switch` instead.

jlebon commented:
Hmm, pkglayering something when in rojig mode gives me:

```
Feb 01 18:50:23 f27-ros.localdomain rpm-ostree[9612]: Unable to update state: Reading deployment 0: GPG verification error: Remote "fahc" not found
```

Looking at the origin, I see:

```
[origin]
baserefspec=fahc:fedora-atomic-host
```

This should be a fully qualified refspec, right?

cgwalters commented:
> baserefspec=fahc:fedora-atomic-host

Ooh yes, that was clearly buggy.  Thinking about this, I realized that the previous design of `refspec=rojig://` was wrong - we can't use `refspec` or libostree may try to interpret it, and it can't for `rojig://`.

So in the fixup :arrow_up: I changed to add a new key `rojig=fahc:fedora-atomic-host`.

jlebon commented:
Nice!
@rh-atomic-bot r+ 7121d58

jlebon commented:
@rh-atomic-bot retry

jlebon commented:
Looks like this is breaking `deploy`. Some interaction with `override-commit` perhaps?

```
PASS: tests/check/test-basic.sh 6 rebase onto newer version
Resolving version '1.0.9'
0 metadata, 0 content objects imported
2 metadata, 0 content objects imported
0 metadata, 0 content objects imported
No change.
ERROR: tests/check/test-basic.sh - too few tests run (expected 26, got 6)
ERROR: tests/check/test-basic.sh - exited with status 77
```

cgwalters commented:
> Some interaction with override-commit perhaps?

Yep, accidentally dropped the bit for that in origin parsing when doing the last fixup.  Rebased :surfing_man: and added new fixup.

jlebon commented:
@rh-atomic-bot r+ 3e0d181
